core/src/subgraph/loader.rs

```diff
@@ -0,0 +1,245 @@
+    Q: GraphQlRunner,
+{
+    pub fn new(store: Arc<S>, link_resolver: Arc<L>, graphql_runner: Arc<Q>) -> Self {
+        Self {
+            store,
+            link_resolver,
+            graphql_runner,
+        }
+    }
+
+    fn dynamic_data_sources_query(
+        &self,
+        deployment: &SubgraphDeploymentId,
+        skip: i32,
+    ) -> Result<Query, Error> {
+        // Obtain the "subgraphs" schema
+        let schema = self.store.subgraph_schema(&SUBGRAPHS_ID)?;
+
+        // Construct a query for the subgraph deployment and all its
+        // dynamic data sources
+        Ok(Query {
+            schema,
+            document: parse_query(
+                r#"
+                query deployment($id: ID!, $skip: Int!) {
+                  subgraphDeployment(id: $id) {
+                    dynamicDataSources(orderBy: id, skip: $skip) {
+                      kind
+                      network
+                      name
+                      source { address abi }
+                      mapping {
+                        kind
+                        apiVersion
+                        language
+                        file
+                        entities
+                        abis { name file }
+                        eventHandlers { event handler }
+                      }
+                      templates {
+                        kind
+                        network
+                        name
+                        source { abi }
+                        mapping {
+                          kind
+                          apiVersion
+                          language
+                          file
+                          entities
+                          abis { name file }
+                          eventHandlers { event handler }
+                        }
+                      }
+                    }
+                  }
+                }
+                "#,
+            )
+            .expect("invalid query for dynamic data sources"),
+            variables: Some(QueryVariables::new(HashMap::from_iter(
+                vec![
+                    (String::from("id"), q::Value::String(deployment.to_string())),
+                    (String::from("skip"), q::Value::Int(skip.into())),
+                ]
+                .into_iter(),
+            ))),
+        })
+    }
+
+    fn query_dynamic_data_sources(
+        &self,
+        deployment: SubgraphDeploymentId,
+        query: Query,
+    ) -> impl Future<Item = QueryResult, Error = Error> + Send {
+        self.graphql_runner
+            .run_query(query)
+            .map_err(move |e| format_err!("Failed to load manifest `{}`: {}", deployment, e))
+    }
+
+    fn parse_data_sources(
+        &self,
+        deployment_id: SubgraphDeploymentId,
+        query_result: QueryResult,
+    ) -> Result<Vec<EthereumContractDataSourceEntity>, Error> {
+        let data = match query_result.data.expect("subgraph deployment not found") {
```

Comment:
I was going try this PR out, but I had set the query complexity limit at 2 million, the data sources query is just over that limit, triggering this panic.

I don't think having to set a limit over 2 million is an issue (at least not one we have to deal with now), but it seems we shouldn't be panicking here.

Reply:
We should however validate a that query complexity satifies a minimum to make sure it doesn't interfere with this, the complexity of that query is 2030100.

Reply:
True, we should fail the subgraph instead.

Reply:
We may want a flag to bypass the complexity, because there's never a guarantee that, if we put in that number as the complexity minimum, we won't forget to keep it up to date in the future.

Reply:
@leodasvacas How about we add
```
    fn run_query_with_complexity(
        &self,
        query: Query,
        max_complexity: Option<u64>,
    ) -> QueryResultFuture {
        let result = execute_query(
            &query,
            QueryExecutionOptions {
                logger: self.logger.clone(),
                resolver: StoreResolver::new(&self.logger, self.store.clone()),
                deadline: GRAPHQL_QUERY_TIMEOUT.map(|t| Instant::now() + t),
                max_complexity: max_complexity,
                max_depth: *GRAPHQL_MAX_DEPTH,
            },
        );
        Box::new(future::ok(result))
    }
```
to `GraphQlRunner` for internal queries?

Reply:
I've added this in the latest commit: https://github.com/graphprotocol/graph-node/pull/832/commits/5cfc3e101b7f059f06d0eaa3bae80dd57c47d375

Reply:
And the subgraph now fails properly if the dynamic data sources query fails, instead of panicking.

Reply:
@Jannis that's smarter, great!